Add flags to specify the kubeconfig path and context

hhiroshell opened the pull request:
## Type of change

- [x] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] Optimization
- [x] Documentation Update

## Description
This PR adds flags to specify the kubeconfig path and context. The target subcommands are as follows:
- init
- health-check
- destroy
- measure

Additionally, this PR includes some small refactorings below:
- Commonize the initialization logic of client config wrote in multiple locations, by adding a new struct KubeClientProvider. This struct provides kubernetes client objects based on single initialization logic.
- Replace `*kubernetes.clientSet` with `kubernetes.Interface` adhere to the concept of utilizing objects through an interface.

## Related Tickets & Documents

- Related Issue #535 
- Closes #535

rsevilla87 commented on pkg/config/config.go:
```diff
@@ -153,20 +150,61 @@ func Parse(uuid string, f io.Reader) (Spec, error) {
+	return &KubeClientProvider{restConfig: restConfig}
+}
+
+func (p *KubeClientProvider) DefaultRestConfig() *rest.Config {
+	restConfig := *p.restConfig
+	return &restConfig
+}
+
+func (p *KubeClientProvider) DefaultClientSet() kubernetes.Interface {
```
These two methods are normally executed together, can you join them into a single one returning *rest.Config and kubernetes.Interface?

hhiroshell replied:
Thank you for your review. Fixed in https://github.com/kube-burner/kube-burner/pull/604/commits/bcdd28f6a78c258951ec679658c9b8f919221a96

rsevilla87 commented on pkg/config/config.go:
```diff
@@ -153,20 +150,61 @@ func Parse(uuid string, f io.Reader) (Spec, error) {
+	return kubernetes.NewForConfigOrDie(p.DefaultRestConfig())
+}
+
+func (p *KubeClientProvider) RestConfig(QPS float32, burst int) *rest.Config {
```
Same as above

hhiroshell replied:
I've done it as well.
https://github.com/kube-burner/kube-burner/pull/604#discussion_r1511165175

vishnuchalla commented on pkg/config/config.go:
```diff
@@ -153,20 +150,53 @@ func Parse(uuid string, f io.Reader) (Spec, error) {
-		kubeconfig = os.Getenv("KUBECONFIG")
-	} else if _, err := os.Stat(filepath.Join(os.Getenv("HOME"), ".kube", "config")); kubeconfig == "" && !os.IsNotExist(err) {
-		kubeconfig = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+type KubeClientProvider struct {
```
I think this one should go into types.go of this package.

hhiroshell replied:
Thank you for your review. Fixed in https://github.com/kube-burner/kube-burner/pull/604/commits/40f952ed81c273a10be07e9a82580b8fde78bb59 .

vishnuchalla commented on pkg/config/config.go:
```diff
@@ -153,20 +150,53 @@ func Parse(uuid string, f io.Reader) (Spec, error) {
+	} else {
+		kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeConfigPath},
+			&clientcmd.ConfigOverrides{CurrentContext: context},
```
What if the user manages to specify the `kubeconfig` and forgets about `context`. Do want to have a default implementation set for that case? For example
```
return clientcmd.BuildConfigFromFlags("", kubeConfigPath)
```
Because it might be a problem when a kubeconfig doesn't have a default context set.

hhiroshell replied:
When `--kubeconfig` is specified, returning the default kubeconfig and context may not align with the user's original intent.
Moreover, this could inadvertently result in executing kube-burner to a cluster that was not intended, which I think is not desirable.

vishnuchalla commented:
In overall, I like this idea. Since this is a new feature, can we get some CI tests added/modified for these cases?
1. With kubeconfig and context provided.
2. Another one with just a kubeconfig input.

hhiroshell commented:
@vishnuchalla 
I've added the test cases you mentioned; please check them.

@vishnuchalla @rsevilla87 
Additionally, I changed the name of the `--kube-config` flag to `--kubeconfig` to align with kubectl and Helm. Note that Helm uses `--kubeconfig`, even though it has `--kube-context` (separated with a hyphen).

vishnuchalla reviewed (approved):
lgtm

hhiroshell commented:
@rsevilla87
I have resolved the conflicts. If there are no issues, please merge it into the main.

rsevilla87 reviewed (approved):
lgtm!, thanks for the PR!
